solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java

```diff
@@ -134,6 +136,7 @@ public SolrDispatchFilter() {
   @Override
   public void init(FilterConfig config) throws ServletException
   {
+    instance = this;
```

Comment:
This does not have to be threadsafe. This is invoked by the Servlet container only once during the life time of this filter

Reply:
But `instance` is static to the class. Not the Filter object instance. Is Filter itself a singleton ?

Reply:
Is there some classloader isolation I'm not following?

Reply:
If there is a singleton instance of this stored somewhere in the servlet container, can we expose a safe static accessor method there? This feels very hacky.

Reply:
> Is Filter itself a singleton?

It's possible to have as many instances of a Filter as they are configured in the `web.xml`. In this case there is only one instance of `SolrDispatchFilter` in the container

Reply:
>Is there some classloader isolation I'm not following?

No. `PrometheusServlet` & `SolrDispatchFilter` are loaded by the same classloader . That's why we are able to share the Object as a static field

Reply:
The other choice is to use 

`getServletContext().setAttribute("some-name", <solr-dispatch-filter-instance>);`

and `PrometheusServlet` can access it using , 
`getServletContext().getAttribute("some-name");`

Reply:
Using the servlet context would be more idiomatic yes! But the prometheus exporter static functions don't have access to any config object to call `getServletContext()`. So they can't read things from there easily.

And while `init()` is called once so there isn't a racing write. The read from the prometheus exporter happens on a different thread, and can in fact race. And I'm not sure if Java is like other systems, but you might do worse than simply read null or not null. On a multicore numa system you can poison the registers with a racing read without synchronization.

I would suggest adding a static getter and setter that use a simple read/write lock around the instance set (Write lock) and instance read (read lock). And add some lifecycle comments around this code so people understand the relationship to the servlet container..

Reply:
Been a while since I did any serious Jetty dev though. Maybe @dragonsinth or @jhump can advise on a more idiomatic pattern for this.

Reply:
>On a multicore numa system you can poison the registers with a racing read without synchronization.

This is a just a single object. It can either be `null` or not `null`. If it's null the worst thing that can happen is that you don't see any cache metrics. That's true, even if you fire a metrics request without ever using the cache . I don't think it can pollute any registry by reading. The metrics values are not kept in any registry. they are read just in time.

Reply:
>This is a just a single object. It can either be null or not null. 

I understand this.

> I don't think it can pollute any registry by reading. The metrics values are not kept in any registry. they are read just in time.

I don't think you understood me. I means CPU register. CPUs have multi level caches. And if you have a thread with a different CPU core affinity doing a write. And another doing a read. Memory writes on one core (depending on the CPU architecture) might not invalidate caches on all other CPU cores. You typically force a memory barrier/synchronization to cause data to be refetched from main memory. 

https://stackoverflow.com/questions/17046558/how-does-a-mutex-ensure-a-variables-value-is-consistent-across-cores

https://www.tutorialspoint.com/parallel_computer_architecture/parallel_computer_architecture_cache_coherence_synchronization.htm

I don't know if we would be subject to this issue or if Java has special mitigations. Where all you need to ensure is ordering. And don't need any memory barriers.

Reply:
Given that reads and writes happen on different threads (stats publisher reads and emits metrics on a different thread from the servlet container initializing the static object). 
And as far as I know, we are not guaranteed to **_not_** read before writing. 

Why not add synchronization or read/write lock here? Even if you don't need it, the lifecycle is not documented at all in any of the read or write points.

Reply:
Let's see a few possible scenarios.

a) PrometheusServlet is invoked in thread 1 while the `SolrDispatchFIlter.init()` is invoked. There is a chance that, SolrDispatchFilter.instance is initialized and the thread does not see because the registers have stale values. It gets the `instance` as null. So, is it any different from the situation where the PrometheusServlet is invoked before `SolrDispatchFIlter.init()` is invoked?
b) `PrometheusServlet` is invoked after the  `SolrDispatchFIlter.init()` is invoked. The `instance` variable is never modified once it is set. It is "effectively" final. So , the registers may have this value cached and it will always be the same value.

I think making the `instance variable `volatile` makes more sense

Reply:
Yup! `volatile` would solve that issue.

Reply:
I'll push those changes

Reply:
But at the expense of performance I think for subsequent reads. But I doubt that would be an issue here given the frequency of stats publishing.

Reply:
>But I doubt that would be an issue here given the frequency of stats publishing.

volatile vs. non-volatile is not perceptibly slower. Moreover, we are talking about stats being collected once in several seconds?

Reply:
The `doFilter` method in the dispatch filter has a check on a volatile variable, otherwise wait on a countdown latch. So there is already memory barrier for the initialization. To leverage this barrier, we can move `instance = this` two lines above and before `this.cores = coresInit`.

However, I think what's better is that since dispatch filter is mapped to the path `*`, it runs for `PrometheusServlet` too, the dispatch filter can expose the `cores` in the request attributes or servlet attributes, then `PrometheusServlet` can retrieve it via the request object, no need to worry about the initialization order, although as mentioned above, I think the initialization is done.

There can be some refactoring like extracting the container stuff in the dispatch filter into a separate class, then the container can be accessed in different places, but could be out of scope for this PR.

Reply:
I really like Shan's suggestion above. Simple and no sharp corners.